### DISCUSSION TITLE
Manually calling "createWithAttributes" does not work because Event is not called

### DISCUSSION THREAD

**Discussion by renepardon** [spatie/laravel-event-sourcing#320] [Q&A]
This is my method within Upload model:

```php
    public static function createWithAttributes(array $attributes): self
    {
        $attributes['id'] = Str::orderedUuid()->toString();

        event(new UploadCreated($attributes));

        return static::uuid($attributes['id']);
    }
```

I try to call it like that within my test and it fails because return value of **createWithAttributes** is null:

```php
    public function test_model_triggers_event_on_creation()
    {
        Event::fake();

        Upload::createWithAttributes($this->getFakeUploadAttributes($this->user, true));

        Event::assertDispatched(UploadCreated::class);
    }
```

BUT this works:

```php
    $aggregate = UploadAggregate::retrieve(self::UPLOAD_UUID)
            ->createUpload($this->getFakeUploadAttributes($user))
            ->persist();

    return Upload::uuid($aggregate->uuid());
```

Is there something I miss? Shouldn't the events be called synchronous and Projector be called immediately which persists the model instance?

```php
    public function onUploadCreated(UploadCreated $event)
    {
        Upload::create($event->attributes);
    }
```

**static::uuid** is just a helper as shown in the documentation. (https://spatie.be/docs/laravel-event-sourcing/v6/using-projectors/writing-your-first-projector)

**Comment by brendt** [comment]
I think your events aren't dispatched because you're calling `Event::fake()`; our event bus hooks into Laravel's event bus, so faking it will simply not trigger anything on the event-sourcing side.

Maybe the [testing aggregates](https://spatie.be/docs/laravel-event-sourcing/v7/using-aggregates/testing-aggregates) section in the docs can help you point in the right direction.

**Reply by renepardon** [reply]
Yes I figured it out, that **Event::fake()** is the problem here. But without calling **Event::fake()** the corresponding assertion **Event::assertDispatched()** is not available as well which makes it a chicken-egg-problem.

Testing the aggregates will not test the Model method itself, sadly. The provided link returns a HTTP 504 right now

**Reply by brendt** [reply]
Have you considered using `Projectionist::fake()` ? https://github.com/spatie/laravel-event-sourcing/blob/main/src/Projectionist.php#L39
